File: May-LeetCoding-Challenge/03-Ransom-Note/Ransom-Note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        if(ransomNote == null || magazine == null){
            return false;
        }
        
        if(ransomNote.length() > magazine.length()) {
            return false;
        }
        /*
        // String solution
        Character tmp;
        for(int i = 0; i < ransomNote.length(); i++) {
            tmp = ransomNote.charAt(i);
            if(magazine.indexOf(tmp) != -1) {
                magazine = magazine.replaceFirst(tmp.toString(), "\0");
            } else {
                return false;
            }
        }
        */
        
        // Linked List solution
        List<Character> magazineLinkedList = new LinkedList<>();
        for(int i = 0; i < magazine.length(); i++) {
            magazineLinkedList.add(magazine.charAt(i));
        }
        
        Character tmp;
        for(int i = 0; i < ransomNote.length(); i++) {
            tmp = ransomNote.charAt(i);
            if(magazineLinkedList.contains(tmp)) {
                // will remove the first occurrence of tmp
                magazineLinkedList.remove(tmp);
            } else {
                return false;
            }
        }
        
        return true;
    }
}

File: May-LeetCoding-Challenge/03-Ransom-Note/Ransom-Note.kt
class RansomNoteKotlin383 {
    fun canConstruct(ransomNote: String, magazine: String): Boolean {
        // 292ms
        if (ransomNote.length > magazine.length) {
            return false
        }
        val intMagazines = IntArray(26)
        // a -> 97
        magazine.forEach { ++intMagazines[it - 'a'] }
        ransomNote.forEach {
            if (--intMagazines[it - 'a'] < 0) {
                return false
            }
        }
        return true
        /*  396ms
        if (ransomNote.length > magazine.length) {
            return false
        }
        val intMagazines = IntArray(26)
        // a -> 97
        magazine.forEach { ++intMagazines[it.toInt() - 97] }
        ransomNote.forEach {
            if (--intMagazines[it.toInt() - 97] < 0) {
                return false
            }
        }
        return true
         */
    }
}